core/src/main/java/io/grpc/internal/KeepAliveManager.java

```diff
@@ -170,8 +171,10 @@ public synchronized void onDataReceived() {
       }
       // schedule a new ping
       state = State.PING_SCHEDULED;
-      pingFuture =
-          scheduler.schedule(sendPing, keepAliveTimeInNanos, TimeUnit.NANOSECONDS);
+      if (pingFuture == null) {
```

Comment:
You can move this condition check `if (pingFuture == null)` up by replacing the condition check `if (state == State.PING_SCHEDULED)` above with it.